### ISSUE TITLE
feat: Use own embeddings easily

### ISSUE THREAD

**Issue by MentalGear** [tantaraio/voy#35]
Voy looks really promising, especially with it's quick retrieval and low footprint.

I'm wondering if there's an easy way to create new records with my own embedding simply as string input ?

Also, my use case is running it locally in the browser, to index photos, and I expect the overall database to be around 1-2 GB. Does voy provide any optimization like sharding to ensure mobile browsers are not overwhelmed (or automatic release of least used items like vector-storage)?

**Comment by DawChihLiou** [comment]
Hey @MentalGear thanks for checking out Voy! Voy doesn't support automatic partitioning or caching at the moment. Voy is designed for limited environments like mobile and edge servers! The current tree-structured index in WebAssemly is able to provide efficient traversal through your data. I think for your use case it'll perform great. 

I really like your idea about automatic sharding. I'll look into it.

**Comment by MentalGear** [comment]
Hey, glad you like the idea!

Indeed Voy is highly efficient in the retrieval!
Here's some resource on sharding from the excellent [RxDB](https://rxdb.info/rx-storage-sharding.html)

I found that overall in-Memory size per tab varies for mobile browsers quite a bit:
For example (latest OS): iPad: 2GB, iPhone X: 1GB. Android is different again.

On-Device Storage:
Privacy and on-device storage has become quite prominent in regard of the new ML generative abilities.

I was wondering if Voy could integrate, besides a normal export function, storage adapters, most preferable with something like OPFS which is fast and supported by all mayor browsers now. (Reference: https://rxdb.info/rx-storage-opfs.html)

**Comment by DawChihLiou** [comment]
What about using web workers? Using a shared worker might give you more freedom from the memory limitation per tab.

**Comment by MentalGear** [comment]
Yes, I'm also interested in this, but could not find definitive resources for web workers. Shared Web workers allow the usage of the same worker across multiple tabs of the same web resource, however I don't think they expand overall available memory (but I don't know for sure).

There's a few APIs, however with very limited browser support (https://developer.mozilla.org/en-US/docs/Web/API/Performance/memory).

**Comment by DawChihLiou** [comment]
Hey @MentalGear I've done some more digging. Because Voy uses the embeddings as coordinates to store and locate minimal data from the resources, the requirement for the storage is greatly reduced. For the resource data, like title, description, and url, I can provide you with an API that allows serialized and unstructured data to further minimize the size.

**Comment by MentalGear** [comment]
Yes, adding custom data to records would be a huge plus !
